legion/engine/core/serialization/serialization.inl

```diff
@@ -0,0 +1,123 @@
+#include <core/serialization/serialization.hpp>
+#pragma once
+
+namespace legion::core::serialization
+{
+    template<typename ViewType, typename Type>
+    inline common::result<void, fs_error> serialize(ViewType& s_view, Type&& data, std::string_view name)
+    {
+        std::vector<std::string> warnings{};
+
+        auto serializer = serializer_registry::get_serializer<remove_cvr_t<Type>>();
+        auto result = serializer->serialize(&data, s_view, name);
+        PropagateErrors(result, warnings);
+
+        return { common::success, warnings };
+    }
+
+    template<typename ViewType, typename Type>
+    inline common::result<void, fs_error> write(fs::view file, Type&& data, std::string_view name)
+    {
+        ViewType view{};
+        std::vector<std::string> warnings{};
+
+        {
+            auto serializer = serializer_registry::get_serializer<remove_cvr_t<Type>>();
+            auto result = serializer->serialize(&data, view, name);
+            PropagateErrors(result, warnings);
+        }
+
+        {
+            auto result = view.write(file);
+            PropagateErrors(result, warnings);
+        }
+
+        return { common::success, warnings };
+    }
+
+    template<typename ViewType, typename Type>
+    inline L_ALWAYS_INLINE common::result<void, fs_error> write(Type&& data, fs::view file, std::string_view name)
+    {
+        return write(file, std::forward<Type>(data), name);
+    }
+
+    template<typename Type, typename ViewType>
+    common::result<Type, fs_error> deserialize(ViewType& s_view, std::string_view name)
+    {
+        byte rawData[sizeof(Type)];
+        Type* data = reinterpret_cast<Type*>(rawData);
+        std::vector<std::string> warnings{};
+
+        auto serializer = serializer_registry::get_serializer<remove_cvr_t<Type>>();
+        auto result = serializer->deserialize(data, s_view, name);
+        PropagateErrors(result, warnings);
+
+        return { *data, warnings };
+    }
+
+    template<typename ViewType, typename Type>
+    inline common::result<Type, fs_error> load(const fs::view& file, std::string_view name)
+    {
+        ViewType view{};
+        byte rawData[sizeof(Type)];
```

Comment:
yeah static_assert would be good, passing pointer types or reference types would break compilation anyways. but at least with a static_assert we could tell you why.